File: taller_hilos/ejercicio1/ejercicio1.py
import threading

# Ejercicio 1: Hilo que imprime números del 1 al 10
#
# En Python no existe la interfaz Runnable; en su lugar se pasa
# una función como argumento 'target' al constructor Thread.
# Esa función hace el papel de run().

def contador():
    # Este código se ejecutará en un hilo separado al del programa principal.
    for i in range(1, 11):
        print(f"Número: {i}")

# Creamos el hilo asociando la función 'contador' como tarea.
hilo = threading.Thread(target=contador)

# start() lanza el hilo. A partir de aquí contador() corre
# concurrentemente con el resto del programa.
hilo.start()

# join() hace que el hilo principal espere a que este hilo termine
# antes de continuar. Sin join(), main podría terminar antes que el hilo.
hilo.join()

File: taller_hilos/ejercicio1/Ejercicio1.java
// Ejercicio 1: Hilo que imprime números del 1 al 10
//
// Implementamos Runnable en lugar de extender Thread para mantener
// la clase libre de heredar comportamiento innecesario de Thread.
// El método run() es el punto de entrada del hilo: todo lo que esté
// aquí se ejecuta cuando el hilo arranca.

class Contador implements Runnable {
    public void run() {
        // Este bucle se ejecuta dentro del hilo secundario,
        // no en el hilo principal (main).
        for (int i = 1; i <= 10; i++) {
            System.out.println("Número: " + i);
        }
    }
}

public class Ejercicio1 {
    public static void main(String[] args) {
        // Envolvemos la tarea (Runnable) dentro de un Thread.
        // El objeto Thread sabe cómo lanzar un hilo del sistema operativo.
        Thread hilo = new Thread(new Contador());

        // start() crea el hilo y llama a run() en ese nuevo hilo.
        // Si llamáramos a hilo.run() directamente, se ejecutaría en main
        // sin ninguna concurrencia.
        hilo.start();
    }
}
